File: examples/strategies/macd_zero_cross/python/strategy.py
"""The MACD zero-cross rule, independent from APIs and order execution."""

from collections.abc import Sequence
from dataclasses import dataclass
from decimal import Decimal
from typing import Literal

FAST_WINDOW = 12
SLOW_WINDOW = 26
SIGNAL_WINDOW = 9

Signal = Literal["entry", "exit"] | None


@dataclass(frozen=True, slots=True)
class Evaluation:
    macd: Decimal
    signal_line: Decimal
    histogram: Decimal
    signal: Signal


def required_closes(
    slow_window: int = SLOW_WINDOW,
    signal_window: int = SIGNAL_WINDOW,
) -> int:
    """Closes needed for a signal line and for three MACD values.

    The first MACD value needs ``slow_window`` closes. The signal line adds
    ``signal_window - 1`` more, and the two-declining-bars rule compares three
    MACD values.
    """

    return max(slow_window + signal_window - 1, slow_window + 2)


# Closes kept while streaming. Ten slow windows make the EMA seed irrelevant to
# the newest value, so a restart converges on the same MACD.
HISTORY_CLOSES = SLOW_WINDOW * 10
MINIMUM_CLOSES = required_closes()


def _ema(values: Sequence[Decimal], window: int) -> list[Decimal]:
    """Return the EMA series, seeded with the average of the first window."""

    multiplier = Decimal(2) / Decimal(window + 1)
    current = sum(values[:window], start=Decimal(0)) / Decimal(window)
    series = [current]
    for value in values[window:]:
        current = (value - current) * multiplier + current
        series.append(current)
    return series


def macd_series(
    closes: Sequence[Decimal],
    fast_window: int = FAST_WINDOW,
    slow_window: int = SLOW_WINDOW,
) -> list[Decimal]:
    """Return MACD values, one per close that both EMAs cover."""

    fast = _ema(closes, fast_window)
    slow = _ema(closes, slow_window)
    aligned = fast[len(fast) - len(slow) :]
    return [f - s for f, s in zip(aligned, slow, strict=True)]


def evaluate(
    closes: Sequence[Decimal],
    in_position: bool = False,
    fast_window: int = FAST_WINDOW,
    slow_window: int = SLOW_WINDOW,
    signal_window: int = SIGNAL_WINDOW,
) -> Evaluation:
    """Return current MACD values and an entry/exit signal.

    Entry is the MACD line crossing zero upwards while flat. Exit is two
    consecutive declining MACD values while holding a position. The signal line
    and histogram are reported for context; neither rule reads them.
    """

    if fast_window <= 0 or slow_window <= fast_window:
        raise ValueError("windows must satisfy 0 < fast_window < slow_window")
    if signal_window <= 0:
        raise ValueError("signal_window must be positive")
    minimum = required_closes(slow_window, signal_window)
    if len(closes) < minimum:
        raise ValueError(f"at least {minimum} closes are required")

    macd = macd_series(closes, fast_window, slow_window)
    signal_line = _ema(macd, signal_window)[-1]
    older, previous, current = macd[-3], macd[-2], macd[-1]

    signal: Signal = None
    if in_position:
        if current < previous < older:
            signal = "exit"
    elif previous <= 0 < current:
        signal = "entry"

    return Evaluation(current, signal_line, current - signal_line, signal)

File: examples/strategies/macd_zero_cross/node/strategy.ts
export const FAST_WINDOW = 12;
export const SLOW_WINDOW = 26;
export const SIGNAL_WINDOW = 9;

export type Signal = "entry" | "exit" | null;

export type Evaluation = Readonly<{
  macd: number;
  signalLine: number;
  histogram: number;
  signal: Signal;
}>;

/**
 * Closes needed for a signal line and for three MACD values.
 *
 * The first MACD value needs `slowWindow` closes. The signal line adds
 * `signalWindow - 1` more, and the two-declining-bars rule compares three MACD
 * values.
 */
export const requiredCloses = (slowWindow = SLOW_WINDOW, signalWindow = SIGNAL_WINDOW): number =>
  Math.max(slowWindow + signalWindow - 1, slowWindow + 2);

/**
 * Closes kept while streaming. Ten slow windows make the EMA seed irrelevant to
 * the newest value, so a restart converges on the same MACD.
 */
export const HISTORY_CLOSES = SLOW_WINDOW * 10;
export const MINIMUM_CLOSES = requiredCloses();

const itemAt = (values: readonly number[], index: number): number => {
  const item = values.at(index);
  if (item === undefined) throw new RangeError(`missing value at index ${index}`);
  return item;
};

/** The EMA series, seeded with the average of the first window. */
const emaSeries = (values: readonly number[], window: number): number[] => {
  const multiplier = 2 / (window + 1);
  let current = values.slice(0, window).reduce((total, item) => total + item, 0) / window;
  const series = [current];
  for (const item of values.slice(window)) {
    current = (item - current) * multiplier + current;
    series.push(current);
  }
  return series;
};

/** MACD values, one per close that both EMAs cover. */
export const macdSeries = (
  closes: readonly number[],
  fastWindow = FAST_WINDOW,
  slowWindow = SLOW_WINDOW,
): number[] => {
  const fast = emaSeries(closes, fastWindow);
  const slow = emaSeries(closes, slowWindow);
  const offset = fast.length - slow.length;
  return slow.map((slowValue, index) => itemAt(fast, index + offset) - slowValue);
};

/**
 * Pure MACD calculation and zero-cross rule, independent from the SDK.
 *
 * Entry is the MACD line crossing zero upwards while flat. Exit is two
 * consecutive declining MACD values while holding a position. The signal line
 * and histogram are reported for context; neither rule reads them.
 */
export const evaluate = (
  closes: readonly number[],
  inPosition = false,
  fastWindow = FAST_WINDOW,
  slowWindow = SLOW_WINDOW,
  signalWindow = SIGNAL_WINDOW,
): Evaluation => {
  if (fastWindow <= 0 || slowWindow <= fastWindow) {
    throw new RangeError("windows must satisfy 0 < fastWindow < slowWindow");
  }
  if (signalWindow <= 0) throw new RangeError("signalWindow must be positive");
  const minimum = requiredCloses(slowWindow, signalWindow);
  if (closes.length < minimum) {
    throw new RangeError(`at least ${minimum} closes are required`);
  }

  const macd = macdSeries(closes, fastWindow, slowWindow);
  const signalLine = itemAt(emaSeries(macd, signalWindow), -1);
  const older = itemAt(macd, -3);
  const previous = itemAt(macd, -2);
  const current = itemAt(macd, -1);

  let signal: Signal = null;
  if (inPosition) {
    if (current < previous && previous < older) signal = "exit";
  } else if (previous <= 0 && current > 0) {
    signal = "entry";
  }

  return { macd: current, signalLine, histogram: current - signalLine, signal };
};
